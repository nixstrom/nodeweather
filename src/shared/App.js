import React, { PureComponent } from 'react'
import { hydrate } from 'react-dom'
import Form from './Form'
import { getWindDirection, formatTemperature } from './utils'

class App extends PureComponent {
  constructor(props) {
    super(props)

    this.state = {
      city: 'hulla'
    }

    let initialState
    // props are not persisted between server and client render, so we store it globally
    if (__isBrowser__) {
      initialState = window.__INITIAL_DATA__
      delete window.__INITIAL_DATA__
    } else {
      initialState = props
    }

    this.state = {
      ...initialState,
      errorMessage: ''
    }

    this.cityInput = React.createRef();
  }

  componentDidMount() {
    window.addEventListener('popstate', this.handleCityChange(), false);
  }

  componentWillUnmount() {
    window.removeEventListener('popstate', this.handleCityChange(), false);
  }

  handleCityChange() {
    const urlParams = new URLSearchParams(window.location.search);
    if (urlParams && urlParams.has('city')) {
      const city = urlParams.get('city')

      fetch(`http://api.openweathermap.org/data/2.5/weather?q=${city},dk&appid=166d00e26d3ff2c6149e89feccc5c59a`)
        .then(response => response.json())
        .then(json => {
          try {
            this.setState({
              city: json.name,
              temperature: json.main.temp,
              humidity: json.main.humidity,
              windSpeed: json.wind.speed,
              windAngle: json.wind.deg,
              errorMessage: ''
            })
          } catch (ex) {
            throw new Error(ex);
          }
      }).catch((ex) => {
        console.log('parsing failed', ex)
        this.setState({ errorMessage: 'Something went wrong. Please try again later.' })
      })
    }
  }

  handleSubmit(e) {
    e.preventDefault()
    const city = this.cityInput.current.value

    history.pushState({}, `ðŸŒ¤ Weather in ${city}`, `/?city=${city}`);

    this.handleCityChange(city)
  }

  render() {
    const {
      city,
      temperature,
      windSpeed,
      windAngle,
      humidity,
      errorMessage,
    } = this.state

    return (
      <div className="widget" style={{ margin: '10px', width: '300px' }}>
        <div className="panel panel-info">
          <div className="panel-heading">Weather in <b>{city}</b></div>
          <ul className="list-group">
            <li className="list-group-item">Temperature: <b>{formatTemperature(temperature)}Â°C</b></li>
            <li className="list-group-item">Humidity: <b>{humidity}%</b></li>
            <li className="list-group-item">Wind: <b>{windSpeed} m/s {getWindDirection(windAngle)}</b></li>
            <li className="list-group-item">
              <Form
                cityInputRef={ this.cityInput }
                onSubmit={ e => this.handleSubmit(e) }
              />
            </li>
            { errorMessage && (<div className="alert alert-danger" role="alert">
                { errorMessage }
              </div>)
            }
          </ul>
        </div>
    </div>
    )
  }
}

export default App
