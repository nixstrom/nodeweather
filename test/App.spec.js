import React from 'react'
import { shallow } from 'enzyme'
//import sinon from 'sinon'

import App from '../src/shared/App'

const initialProps = {
  city: 'Copenhagen',

}

describe('<App />', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(<App
        city="Copenhagen"
        temp={ 276 }
        humidity={ 80 }
        windSpeed={ 5.1 }
        windAngle={ 180 }
      />)
  })
  it('renders App', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders App with errorMessage', () => {
    wrapper.setState({ errorMessage: 'An error occured' })
    wrapper.update()

    // in theory one of these two ass is redundant, but as there are opinions
    // for and against each approach, I am demonstrating both
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.alert.alert-danger').length).toBe(1)
  })
})
