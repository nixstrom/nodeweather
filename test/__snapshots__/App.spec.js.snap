// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders App 1`] = `
<div
  className="widget"
  style={
    Object {
      "margin": "10px",
      "width": "300px",
    }
  }
>
  <div
    className="panel panel-info"
  >
    <div
      className="panel-heading"
    >
      Weather in 
      <b>
        Copenhagen
      </b>
    </div>
    <ul
      className="list-group"
    >
      <li
        className="list-group-item"
      >
        Temperature: 
        <b>
          3
          °C
        </b>
      </li>
      <li
        className="list-group-item"
      >
        Humidity: 
        <b>
          80
          %
        </b>
      </li>
      <li
        className="list-group-item"
      >
        Wind: 
        <b>
          5.1
           m/s 
          S
        </b>
      </li>
      <li
        className="list-group-item"
      >
        <Form
          cityInputRef={
            Object {
              "current": null,
            }
          }
          onSubmit={[Function]}
        />
      </li>
    </ul>
  </div>
</div>
`;

exports[`<App /> renders App with errorMessage 1`] = `
<div
  className="widget"
  style={
    Object {
      "margin": "10px",
      "width": "300px",
    }
  }
>
  <div
    className="panel panel-info"
  >
    <div
      className="panel-heading"
    >
      Weather in 
      <b />
    </div>
    <ul
      className="list-group"
    >
      <li
        className="list-group-item"
      >
        Temperature: 
        <b>
          NaN
          °C
        </b>
      </li>
      <li
        className="list-group-item"
      >
        Humidity: 
        <b>
          %
        </b>
      </li>
      <li
        className="list-group-item"
      >
        Wind: 
        <b>
           m/s 
        </b>
      </li>
      <li
        className="list-group-item"
      >
        <Form
          cityInputRef={
            Object {
              "current": null,
            }
          }
          onSubmit={[Function]}
        />
      </li>
      <div
        className="alert alert-danger"
        role="alert"
      >
        An error occured
      </div>
    </ul>
  </div>
</div>
`;
